### PR TITLE
[WIP] Presentation header

### PR DESCRIPTION
**Pull request by quartzjer**

### CODE REVIEW

File: draft-jmiller-json-web-proof.md [hunk 1]
```diff
@@ -46,15 +46,15 @@ organization = "Microsoft"
...
 
-This document defines a new container format similar in purpose and design to JSON Web Signature (JWS) called a _JSON Web Proof (JWP)_.  Unlike JWS, which integrity-protects only a single payload, JWP can integrity-protect multiple payloads, enabling those payloads to be selectively disclosed. It also specifies a proof-generation step, enabling privacy-preserving selection and computation.
+This document defines a new container format similar in purpose and design to JSON Web Signature (JWS) called a _JSON Web Proof (JWP)_.  Unlike JWS, which integrity-protects only a single payload, JWP can integrity-protect multiple payloads in one message.  It also specifies a new presentation form that supports selective disclosure of individual payloads, enables additional proof computation, and adds a protected header to prevent replay and support binding mechanisms.
```
Comment (OR13) [inline draft-jmiller-json-web-proof.md:49]: this is the key change.

File: draft-jmiller-json-web-proof.md [hunk 2]
```diff
@@ -67,29 +67,47 @@ There are a growing number of these cryptographic primitives that support select
 * [U-Prove](https://www.microsoft.com/en-us/research/project/u-prove/)
 * [Spartan](https://github.com/microsoft/Spartan)
 
-All of these follow the same pattern of taking multiple claims (a.k.a., "attributes" or "messages" in the literature) and binding them together into an issued token.  These are then later securely one-way transformed into a presentation that reveals potentially only a subset of the original claims, predicate proofs of the claim values, or proofs of knowledge of the claims.
+All of these follow the same pattern of taking multiple claims (a.k.a., "attributes" or "messages" in the literature) and binding them together into a single issued statement.  These are then later securely one-way transformed into a presentation that reveals potentially only a subset of the original claims, predicate proofs about the claim values, or proofs of knowledge of the claims.
```
Comment (selfissued) [inline draft-jmiller-json-web-proof.md:70]: "statement" reads oddly.  Please change it back to "token".

File: draft-jmiller-json-web-proof.md [hunk 3]
```diff
@@ -67,29 +67,47 @@ There are a growing number of these cryptographic primitives that support select
...
 
 # Conventions and Definitions
 
-The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD",
-"SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and "OPTIONAL" in this
-document are to be interpreted as described in BCP 14 [@!RFC2119] [@RFC8174]
-when, and only when, they appear in all capitals, as shown here.
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in BCP 14 [@!RFC2119] [@RFC8174] when, and only when, they appear in all capitals, as shown here.
+
+The roles of "issuer", "holder", and "verifier", are used from and defined by the [Verifiable Credentials Data Model v1.1](https://www.w3.org/TR/2021/REC-vc-data-model-20211109/).  The term "presentation" is also used and defined by this source, but the term "credential" is avoided in this specification in order to minimize confusion with other definitions.
```
Comment (selfissued) [inline draft-jmiller-json-web-proof.md:77]: used from and defined by -> used as defined by (twice)

File: draft-jmiller-json-web-proof.md [hunk 4]
```diff
@@ -67,29 +67,47 @@ There are a growing number of these cryptographic primitives that support select
...
 
 # Background
 
 A _JSON Web Proof (JWP)_ is very similar to a JWS [@RFC7515], with the addition that it can contain multiple individual secured payloads instead of a singular one.  JWP-supporting algorithms are then able to separate and act on the individual payloads contained within.
 
-In addition to the JWS `sign` and `verify` interactions, JWP also importantly adds a `prove` processing step for interaction with the algorithm to perform the selective disclosure and privacy-preserving transformations.  This allows for multi-party interactions where a token is issued from one party, held by another party, then used to generate and present proofs about the token to another verifying party.  While `sign` only occurs when a JWP is being created, the `prove` and `verify` steps may be safely repeated on a signed JWP (when supported by the algorithm).
+The intent of JSON Web Proofs is to establish a common container format for multiple payloads that can be integrity-verified against a cryptographic proof value also in the container.  It does not create or specify any cryptographic protocols, multi-party protocols, or detail any algorithm specific capabilities.
```
Comment (selfissued) [inline draft-jmiller-json-web-proof.md:90]: algorithm specific -> algorithm-specific

File: draft-jmiller-json-web-proof.md [hunk 5]
```diff
@@ -67,29 +67,47 @@ There are a growing number of these cryptographic primitives that support select
...
+
+In order to fully support the newer privacy primitives, JWP introduces the three roles of issuer, holder, and verifier as defined by [@VCDM11].  There are also two forms of a JWP: the signed form created by an issuer for a holder, and the presented form created by a holder for a verifier.
+
+A JWP is initially created by the issuer using a `sign` interaction with an implementation.  A successful results is a signed JWP that has a single issuer-protected header, one or more payloads, and an initial proof value that contains the signing algorithm output.  The holder upon receiving a signed JWP then uses `validate` to check the integrity protection of the header and all payloads using the given proof value.
```
Comment (selfissued) [inline draft-jmiller-json-web-proof.md:94]: results -> result
Add commas around "upon receiving a signed JWP".

File: draft-jmiller-json-web-proof.md [hunk 6]
```diff
@@ -67,29 +67,47 @@ There are a growing number of these cryptographic primitives that support select
...
 
-The intent of JSON Web Proofs is to establish a common container format for multiple payloads that can be integrity-verified against a cryptographic proof value also in the container.  It does not create or specify any cryptographic protocols, interaction protocols, or custom options for algorithms with additional capabilities.
+After validation, the holder uses `present` to apply any selective disclosure choices, perform privacy-preserving transformations for unlinkability, and add a presentation-protected header that ensures the resulting presented JWP cannot be replayed.  The verifier then uses `verify` to ensure the integrity protection of the protected headers and any disclosed payloads along with verifying any additional ZKPs covering non-disclosed payloads.
```
Comment (selfissued) [inline draft-jmiller-json-web-proof.md:96]: Add a comma after "disclosed payloads".

File: draft-jmiller-json-web-proof.md [hunk 7]
```diff
@@ -67,29 +67,47 @@ There are a growing number of these cryptographic primitives that support select
...
+
+While `sign` and `validate` only occur when a JWP is initially created by the issuer, the `present` and `verify` steps may be safely repeated by a holder on a signed JWP.  The unlinkability of the resulting presented JWP is only protected when supported by the underlying algorithm.
```
Comment (selfissued) [inline draft-jmiller-json-web-proof.md:98]: protected -> provided

File: draft-jmiller-json-web-proof.md [hunk 8]
```diff
@@ -67,29 +67,47 @@ There are a growing number of these cryptographic primitives that support select
...
 
 Algorithm definitions that support JWPs are being done in separate companion specifications - just as the [JSON Web Algorithms] [@RFC7518] specification does for JWS and JWE [@RFC7516].  The JSON Proof Algorithms specification defines how an initial set of algorithms are used with JWP.
 
 # JWP Format
 
-A JWP always contains a protected header along with one or more payloads.  Each payload is represents an octet string.  The payloads are then serialized as an ordered array.  Finally, the JWP contains a representation of a cryptographic proof over the payloads.
+A JWP may have one or two integrity protected headers depending on if it is the signed form (one) or the presented form (two).  The protected headers are always a JSON object that is serialized as a UTF-8 encoded octet string.
```
Comment (selfissued) [inline draft-jmiller-json-web-proof.md:104]: Add a comma after "integrity protected headers".

File: draft-jmiller-json-web-proof.md [hunk 9]
```diff
@@ -67,29 +67,47 @@ There are a growing number of these cryptographic primitives that support select
...
+
+Every JWP has one or more payloads, each payload is an octet string.  All of the contained payloads are serialized as an array who's ordering is preserved in all serializations.
```
Comment (selfissued) [inline draft-jmiller-json-web-proof.md:106]: , -> ;
who's -> for which the

File: draft-jmiller-json-web-proof.md [hunk 10]
```diff
@@ -67,29 +67,47 @@ There are a growing number of these cryptographic primitives that support select
...
+
+The JWP proof value is a single octet string that is only generated from and consumed by the underling JPA.  Internally the proof value may one or more cryptographic statements that are used to check the integrity protection of the issuer header and all payloads.  Each of these statements may be a ZKP or a traditional cryptographic signature, the algorithm is responsible for how these statements are serialized into a single proof value.
```
Comment (selfissued) [inline draft-jmiller-json-web-proof.md:108]: Add a comma after "Internally".
Break the last sentence into two sentences at the comma.

File: draft-jmiller-json-web-proof.md [hunk 11]
```diff
@@ -67,29 +67,47 @@ There are a growing number of these cryptographic primitives that support select
...
+
+When a signed JWP is presented it undergoes a transformation that adds a presentation protected header. It may also have one or more payloads hidden, disclosing a subset of the original signed payloads.  The proof value will always be updated to add integrity protection of the presentation header along with the necessary cryptographic statements to verify the presented JWP.
```
Comment (selfissued) [inline draft-jmiller-json-web-proof.md:110]: Add a comma after "presented".

**Review by selfissued — APPROVED** [review]
Please address the nits in this review, then merge!

File: draft-jmiller-json-web-proof.md [hunk 12]
```diff
@@ -46,15 +46,15 @@ organization = "Microsoft"
...
 
-With these new use cases, there is an increased focus on adopting privacy-protecting cryptographic primitives.  While such primitives are still an active area of academic and applied research, the leading candidates introduce new patterns that are not currently supported by JOSE.  These new patterns are largely focused on two areas: supporting selective disclosure when presenting information and minimizing correlation through the use of Zero-Knowledge Proofs (ZKPs), instead of traditional signatures.
+With these new use cases, there is an increased focus on adopting privacy-protecting cryptographic primitives.  While such primitives are still an active area of academic and applied research, the leading candidates introduce new patterns that are not currently supported by JOSE.  These new patterns are largely focused on two areas: supporting selective disclosure when presenting information, and minimizing correlation through the use of Zero-Knowledge Proofs (ZKPs) instead of traditional signatures.
```
Comment (dwaite) [inline draft-jmiller-json-web-proof.md:57]: instead of -> in addition to

File: draft-jmiller-json-web-proof.md [hunk 13]
```diff
@@ -67,35 +67,57 @@ There are a growing number of these cryptographic primitives that support select
...
+A JWP is initially created by the issuer using the `issue` interaction with an implementation.  A successful result is an issued JWP that has a single issuer-protected header, one or more payloads, and an initial proof value that contains the issuing algorithm output.  The holder, upon receiving an issued JWP, then uses `validate` to check the integrity protection of the header and all payloads using the given proof value.
+
+After validation, the holder uses `present` to apply any selective disclosure choices, perform privacy-preserving transformations for unlinkability, and add a presentation-protected header that ensures the resulting presented JWP cannot be replayed.  The verifier then uses `verify` to ensure the integrity protection of the protected headers and any disclosed payloads, along with verifying any additional ZKPs covering non-disclosed payloads.
```
Comment (dwaite) [inline draft-jmiller-json-web-proof.md:96]: Would love to suggest better words than 'validate' and 'verify' - perhaps verify becomes 'confirm' a presentation?

File: draft-jmiller-json-web-proof.md [hunk 14]
```diff
@@ -104,50 +126,75 @@ For example:
...
+A JWP may also be single-use, where an issued JWP can only be used once to generate a presented form, any additional presentations would be inherently correlatable.  These are still useful for applications needing only selective disclosure or where new unique issued JWPs can be retrieved easily.
 
-Implementations will need to provide optional arguments for each payload as input into the `prove` step.  These arguments can be used for generating predicate proofs, linking options, etc.
+### Presentation Protected Header
 
-Algorithms SHOULD generate a new un-correlatable proof value during the `prove` step.  A JWP may also be single-use, where correlation across multiple derivations is not a factor.
+The presented form of a JWP MUST contain a presentation protected header.  It is added by the holder and MUST be integrity protected by the underling JPA.
+
+This header is used to ensure a presented JWP can not be replayed and is cryptographically bound to the issuer it was presented to.
```
Comment (dwaite) [inline draft-jmiller-json-web-proof.md:153]: Should this be 'bound to the holder it was presented by'?

**Review by dwaite — CHANGES_REQUESTED** [review]

File: draft-jmiller-json-web-proof.md (hunk 13 above)
Comment (quartzjer) [reply draft-jmiller-json-web-proof.md:96]: s/validate/confirm/

**Review by dwaite — APPROVED** [review]

**Pull request merged**